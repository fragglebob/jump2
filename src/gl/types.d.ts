import { m4 } from "twgl.js";

interface RenderContext {
    world: m4.Mat4;
    viewProjection: m4.Mat4;
}