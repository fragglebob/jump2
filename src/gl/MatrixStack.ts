import { m4, v3 } from "twgl.js"

export class MatrixStack {

    _stack: m4.Mat4[];
    _current: number;

    constructor() {
        this._stack = [m4.identity()]
        this._current = 0;
    }

    reset() {
        this._stack = [m4.identity()]
        this._current = 0;
    }

    public get current() : m4.Mat4 {
        return this._stack[this._current]
    }  

    pop() {
        if(this._current > 0) {
            this._stack.pop();
            this._current--;
        }
    }

    push() {
        this._stack.push(this.copy());
        this._current++;
    }

    copy() : m4.Mat4 {
        return m4.copy(this.current);
    }

    scale(vec: v3.Vec3) : void {
        m4.scale(this.current, vec, this.current)
    }

    translate(vec: v3.Vec3) : void {
        m4.translate(this.current, vec, this.current);
    }

    rotateX(angle: number) : void {
        m4.rotateX(this.current, angle, this.current);
    }

    rotateY(angle: number) : void {
        m4.rotateY(this.current, angle, this.current);
    }

    rotateZ(angle: number) : void {
        m4.rotateZ(this.current, angle, this.current);
    }
}