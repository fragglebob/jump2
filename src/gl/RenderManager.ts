import { m4 } from "twgl.js";
import { MatrixStack } from "./MatrixStack";
import type { Renderer } from "./Renderer";
import type { RenderContext } from "./types";
import { hslToRgb } from "./utils/hsl2rgb";

interface RenderManagerInterface {
  time(): number;
  frame(): number;

  beat(): number;
  bar(): number;
  beat_progress(): number;
  bar_progress(): number;

  beat_raw(): number;

  fft(index: number): number;

  knob(index: number): number;
  slider(index: number): number;

  pushMatrix(): void;
  popMatrix(): void;

  translate(x: number, y: number, z: number): void;

  rotateX(amount: number): void;
  rotateY(amount: number): void;
  rotateZ(amount: number): void;

  scale(amount: number): void;
  scale(x: number, y: number, z: number): void;

  box(): void;
  ball(): void;

  rgb(r: number, g: number, b: number): void;
  rgba(r: number, g: number, b: number, a: number): void;
  hsl(h: number, s: number, l: number): void;
  hsla(h: number, s: number, l: number, a: number): void;

  fx_kale(segments: number): void;
  fx_grid(rows: number): void;
  fx_px(x: number, y?: number): void;
  fx_rgb(amount: number, angle?: number): void;
  fx_bloom(): void;
  fx_feedback(): void;
  fx_warp(size?: number, speed?: number, amount?: number): void;
  fx_ascii(scale?: number): void;
}

type CameraMatrixes = {
  camera: m4.Mat4;
  projection: m4.Mat4;
  view: m4.Mat4;
};

export class RenderManager implements RenderManagerInterface {
  private worldMatrix: MatrixStack;
  private renderer: Renderer;
  private cameraDetails: CameraMatrixes;

  constructor(renderer: Renderer) {
    this.renderer = renderer;
    this.worldMatrix = new MatrixStack();
    this.cameraDetails = this.setupCamera();

    this.renderer.globalUniforms.u_view = this.cameraDetails.view;
    this.renderer.globalUniforms.u_projection = this.cameraDetails.projection;
  }

  private setupCamera(): CameraMatrixes {
    const fov = (30 * Math.PI) / 180;
    const aspect =
      this.renderer.canvas.clientWidth / this.renderer.canvas.clientHeight;
    const zNear = 0.5;
    const zFar = 100;
    const projection = m4.perspective(fov, aspect, zNear, zFar);

    const eye = [0, 0, -15];
    const target = [0, 0, 0];
    const up = [0, 1, 0];
    const camera = m4.lookAt(eye, target, up);

    const view = m4.inverse(camera);

    return {
      camera,
      projection,
      view,
    };
  }

  private createRenderContext(): RenderContext {
    return {
      world: this.worldMatrix.current,
    };
  }

  time() {
    return this.renderer.getTime();
  }

  frame() {
    return this.renderer.getFrame();
  }

  beat(): number {
    return this.renderer.tempoData?.currentBeat ?? 0;
  }

  bar(): number {
    return this.renderer.tempoData?.currentBar ?? 0;
  }

  beat_progress(): number {
    return this.renderer.tempoData?.beatProgress ?? 0;
  }

  bar_progress(): number {
    return this.renderer.tempoData?.barProgress ?? 0;
  }

  beat_raw(): number {
    return this.renderer.tempoData?.beatTotal ?? 0;
  }

  fft(index = 0): number {
    if (
      !this.renderer.fftData ||
      typeof this.renderer.fftData[index] === "undefined"
    ) {
      return 0;
    }
    return this.renderer.fftData[index];
  }

  knob(index: number): number {
    return this.renderer.midiMix.knobs[index] ?? 0;
  }

  slider(index: number): number {
    return this.renderer.midiMix.sliders[index] ?? 0;
  }

	button(name: string): number {
		return this.renderer.gamepad.getButton(name);
	}

	axis(name: string): number {
		return this.renderer.gamepad.getAxis(name);
	}

  rgb(r: number, g: number, b: number) {
    this.rgba(r, g, b, 1);
  }

  rgba(r: number, g: number, b: number, a: number) {
    this.renderer.updateColor([r, g, b, a]);
  }

  hsl(h: number, s: number, l: number) {
    const [r, g, b] = hslToRgb(h, s, l);
    this.rgba(r, g, b, 1);
  }

  hsla(h: number, s: number, l: number, a: number) {
    const [r, g, b] = hslToRgb(h, s, l);
    this.rgba(r, g, b, a);
  }

  pushMatrix() {
    this.worldMatrix.push();
  }

  popMatrix() {
    this.worldMatrix.pop();
  }

  box() {
    this.renderer.renderBox(this.createRenderContext());
  }

  ball() {
    this.renderer.renderBall(this.createRenderContext());
  }

  translate(x: number, y: number, z: number) {
    this.worldMatrix.translate([x, y, z]);
  }

  rotateX(amount: number) {
    this.worldMatrix.rotateX(amount);
  }
  rotateY(amount: number) {
    this.worldMatrix.rotateY(amount);
  }
  rotateZ(amount: number) {
    this.worldMatrix.rotateZ(amount);
  }

  scale(x: number, y?: number, z?: number) {
    if (y != null && z != null) {
      this.worldMatrix.scale([x, y, z]);
    } else {
      this.worldMatrix.scale([x, x, x]);
    }
  }

  fx_kale(segments?: number, time?: number) {
    this.renderer.doRenderPass(this.renderer.passes.kaleidoscope, {
      segments: segments ?? 2,
      time,
    });
  }

  fx_grid(rows: number) {
    this.renderer.doRenderPass(this.renderer.passes.grid, { rows: rows ?? 2 });
  }

  fx_px(x: number, y?: number) {
    this.renderer.doRenderPass(this.renderer.passes.px, {
      x: x ?? 100,
      y: y ?? x ?? 100,
    });
  }

  fx_rgb(amount: number, angle?: number) {
    this.renderer.doRenderPass(this.renderer.passes.rgb, {
      amount: amount ?? 0.01,
      angle,
    });
  }

  fx_bloom() {
    this.renderer.doRenderPass(this.renderer.passes.bloom, {});
  }

  fx_feedback() {
    this.renderer.doRenderPass(this.renderer.passes.feedback, {});
  }

  fx_warp(size?: number, speed?: number, amount?: number, time?: number) {
    this.renderer.doRenderPass(this.renderer.passes.warp, {
      size,
      speed,
      amount,
      time,
    });
  }

  fx_ascii(scale?: number) {
    this.renderer.doRenderPass(this.renderer.passes.ascii, { scale });
  }
}
