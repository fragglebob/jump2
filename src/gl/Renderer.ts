import {
  bindFramebufferInfo,
  BufferInfo,
  createBufferInfoFromArrays,
  createFramebufferInfo,
  createProgramInfo,
  createTexture,
  createVertexArrayInfo,
  drawBufferInfo,
  FramebufferInfo,
  m4,
  primitives,
  ProgramInfo,
  resizeCanvasToDisplaySize,
  resizeFramebufferInfo,
  setBuffersAndAttributes,
  setUniforms,
  VertexArrayInfo,
} from "twgl.js";


import basicVert from "./shaders/basic.vert.glsl";
import basicFrag from "./shaders/basic.frag.glsl";
import screenVert from "./shaders/screen.vert.glsl";
import screenFrag from "./shaders/screen.frag.glsl";
import { RenderContext } from "./types";
import { RenderManager } from "./RenderManager";
import { KaleidoscopePass } from "./postfx/kaleidoscope/KaleidoscopePass";
import { RenderPass } from "./postfx/RenderPass";


type Vec4 = [number, number, number, number];
type Vec3 = [number, number, number];

export class Renderer {
  readonly gl: WebGL2RenderingContext;

  mainProgramInfo: ProgramInfo;
  screenProgramInfo: ProgramInfo;

  basicTexture: WebGLTexture;

  globalUniforms: {
    u_lightWorldPos: Vec3,
    u_lightColor: Vec4,
    u_ambient: Vec4,
    u_specular: Vec4,
    u_shininess: number,
    u_specularFactor: number,
    u_view: m4.Mat4,
    u_projection: m4.Mat4
  };
  styleUniforms: {
    // u_diffuse: WebGLTexture,
    u_colorMult: Vec4
  };

  objectUniforms: {
    u_world: m4.Mat4
  };

  primatives: {
    rect: VertexArrayInfo,
    ball: VertexArrayInfo,
    box: VertexArrayInfo
  }

  frame: number = 0;
  time: number = 0;

  passes: {
    kaleidoscope: KaleidoscopePass,
  }

  screenBufferInfo: BufferInfo;

  framebuffers: [FramebufferInfo, FramebufferInfo];
  currentFramebufferIndex: number = 0;

  constructor(gl: WebGL2RenderingContext) {
    this.gl = gl;

    this.mainProgramInfo = this.createMainProgram();
    this.screenProgramInfo = this.createScreenProgram();

    this.basicTexture = this.createBasicTexture();

    this.globalUniforms = {
      u_lightWorldPos: [1, 8, -10],
      u_lightColor: [0.8, 0.8, 0.8, 1],
      u_ambient: [0.2, 0.2, 0.2, 1],
      u_specular: [1, 1, 1, 1],
      u_shininess: 50,
      u_specularFactor: 1,
      u_view: m4.identity(),
      u_projection: m4.identity(),
    };

    this.styleUniforms = {
      // u_diffuse: this.basicTexture,
      u_colorMult: [1, 1, 1, 1],
    }

    this.objectUniforms = {
      u_world: m4.identity(),
    };

    this.primatives = {
        box: this.createCube(),
        ball: this.createBall(),
        rect: this.createRect()
    }

    this.screenBufferInfo = primitives.createXYQuadBufferInfo(this.gl, 2);

    const attachments = [
      {
        format: this.gl.RGBA,
        type: this.gl.UNSIGNED_BYTE,
        min: this.gl.LINEAR,
        wrap: this.gl.CLAMP_TO_EDGE
      }, {
        format: this.gl.DEPTH_STENCIL
      }
    ]

    this.framebuffers = [
      createFramebufferInfo(this.gl, attachments),
      createFramebufferInfo(this.gl, attachments),
    ];

    this.passes = {
      kaleidoscope: new KaleidoscopePass(this),
    };
  }

  private createMainProgram(): ProgramInfo {
    return createProgramInfo(this.gl, [basicVert, basicFrag]);
  }

  private createScreenProgram(): ProgramInfo {
    return createProgramInfo(this.gl, [screenVert, screenFrag]);
  }

  private createBasicTexture(): WebGLTexture {
    return createTexture(this.gl, {
      min: this.gl.NEAREST,
      mag: this.gl.NEAREST,
      src: [
        255, 255, 255, 255, 192, 192, 192, 255, 192, 192, 192, 255, 255, 255,
        255, 255,
      ],
    });
  }


  createRect(): VertexArrayInfo {
    return createVertexArrayInfo(this.gl, this.mainProgramInfo, primitives.createPlaneBufferInfo(this.gl, 1, 1));
  }

  createCube(): VertexArrayInfo {
    return createVertexArrayInfo(this.gl, this.mainProgramInfo, primitives.createCubeBufferInfo(this.gl, 1));
  }

  createBall() : VertexArrayInfo {
    return createVertexArrayInfo(this.gl, this.mainProgramInfo, primitives.createSphereBufferInfo(this.gl, 1, 10, 10));
  }

  renderVertexArrayInfo(ctx: RenderContext, vertexArrayInfo: VertexArrayInfo) {
    this.objectUniforms.u_world = ctx.world;
    setUniforms(this.mainProgramInfo, this.objectUniforms);

    setBuffersAndAttributes(this.gl, this.mainProgramInfo, vertexArrayInfo);
    drawBufferInfo(this.gl, vertexArrayInfo);
  }

  renderBox(ctx: RenderContext) {
    this.renderVertexArrayInfo(ctx, this.primatives.box);
  }

  renderBall(ctx: RenderContext) {
    this.renderVertexArrayInfo(ctx, this.primatives.ball);
  }

  update(time: number) : void {
    this.time = time;
    this.frame++;
  }

  getTime() : number {
    return this.time;
  }

  getFrame() : number {
    return this.frame;
  }

  useMainProgram() {
    this.gl.enable(this.gl.DEPTH_TEST);
    this.gl.enable(this.gl.CULL_FACE);
    this.gl.useProgram(this.mainProgramInfo.program);
    this.gl.clear(this.gl.DEPTH_BUFFER_BIT);
  }

  render(callback: (ctx: RenderManager) => void) {

    if(resizeCanvasToDisplaySize(this.gl.canvas)) {
      resizeFramebufferInfo(this.gl, this.framebuffers[0]);
      resizeFramebufferInfo(this.gl, this.framebuffers[1]);
    }

    bindFramebufferInfo(this.gl, this.currentFramebuffer());

    this.gl.clear(this.gl.COLOR_BUFFER_BIT | this.gl.DEPTH_BUFFER_BIT);

    this.useMainProgram();

    setUniforms(this.mainProgramInfo, this.globalUniforms);
    setUniforms(this.mainProgramInfo, this.styleUniforms);

    const manager = new RenderManager(this);

    callback(manager);

    this.gl.useProgram(this.screenProgramInfo.program);
    this.drawFramebuffer(
      this.screenProgramInfo,
      this.currentFramebuffer(),
      null
    )
  }

  renderRenderPass(pass: RenderPass<any>) {
    const currentFramebuffer = this.currentFramebuffer();
    const nextFramebuffer = this.nextFramebuffer();

    this.drawFramebuffer(
      pass.getProgramInfo(),
      currentFramebuffer,
      nextFramebuffer
    );
  }

  currentFramebuffer() : FramebufferInfo {
    return this.framebuffers[this.currentFramebufferIndex];;
  }

  nextFramebuffer() : FramebufferInfo {
    this.currentFramebufferIndex++;
    if(this.currentFramebufferIndex >= this.framebuffers.length) {
      this.currentFramebufferIndex = 0;
    }
    return this.framebuffers[this.currentFramebufferIndex];
  }

  drawFramebuffer(program: ProgramInfo, from: FramebufferInfo, to: FramebufferInfo | null) {
    bindFramebufferInfo(this.gl, to);

    this.gl.clearColor(0, 0, 0, 0);
    this.gl.clear(this.gl.COLOR_BUFFER_BIT | this.gl.DEPTH_BUFFER_BIT);

    setUniforms(program, {
      u_texture: from.attachments[0]
    });

    this.gl.bindVertexArray(null)
    
    setBuffersAndAttributes(this.gl, program, this.screenBufferInfo);
    drawBufferInfo(this.gl, this.screenBufferInfo);
  }

  updateColor(color: Vec4) {
    this.styleUniforms.u_colorMult = color;
    setUniforms(this.mainProgramInfo, this.styleUniforms);
  }

}
