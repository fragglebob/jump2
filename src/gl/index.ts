import type { Analyser } from "../audio/Analyser";
import {
  type UserRenderFunction,
  createFunction,
} from "../compiler/createFunction";
import type { MIDIMix } from "../midi/MIDIMix";
import { Renderer } from "./Renderer";
import {
  type CursorSelection,
  getCursorSelection,
  isCursorSelection,
} from "./utils/getCursorSelection";
import { getStartingCode } from "./utils/getStartingCode";
import { codeStorageKey, updateCodeInLocalStorage } from "./utils/localStorage";

export type AppMode = "editor" | "demo" | "normal";

interface CodeDisplay {
  setIsError(isError: boolean): void;
  setSelection(selection: CursorSelection): void;
  init(): void;
  destory(): void;
  displayCode(code: string): void;
}

class TextareaCodeDisplay implements CodeDisplay {
  readonly textarea: HTMLTextAreaElement;

  app: GLApp;

  constructor(app: GLApp) {
    this.app = app;
    this.textarea = document.createElement("textarea");
  }

  setIsError(isError: boolean): void {
    if (isError) {
      this.textarea.classList.add("error");
    } else {
      this.textarea.classList.remove("error");
    }
  }

  handleTextareaUpdate = () => {
    const code = this.textarea.value;
    this.app.updateCode(code);
    this.sendSelectionUpdate();
  };

  displayCode(code: string) {
    this.textarea.value = code;
  }

  sendSelectionUpdate() {
    const selection = getCursorSelection(this.textarea);
    if (selection) {
      this.app.updateSelection(selection);
    }
  }

  handleTextareaSelectionChannge = () => {
    if (!document.hasFocus()) {
      return;
    }
    this.sendSelectionUpdate();
  };

  init = () => {
    this.app.root.appendChild(this.textarea);
    this.textarea.addEventListener("input", this.handleTextareaUpdate);
    this.textarea.addEventListener(
      "selectionchange",
      this.handleTextareaSelectionChannge,
    );
  };

  destory = () => {
    this.app.root.removeChild(this.textarea);
    this.textarea.removeEventListener("input", this.handleTextareaUpdate);
    this.textarea.removeEventListener(
      "selectionchange",
      this.handleTextareaSelectionChannge,
    );
  };

  setSelection(selection: CursorSelection): void {
    this.textarea.selectionStart = selection.start;
    this.textarea.selectionEnd = selection.end;
  }
}

class DemoCodeDisplay implements CodeDisplay {
  readonly pre: HTMLPreElement;

  app: GLApp;

  code: string;

  selection?: CursorSelection;

  constructor(app: GLApp) {
    this.app = app;
    this.pre = document.createElement("pre");
    this.code = "";
  }
  setIsError(isError: boolean): void {
    if (isError) {
      this.pre.classList.add("error");
    } else {
      this.pre.classList.remove("error");
    }
  }
  setSelection(selection: CursorSelection): void {
    this.selection = selection;
    this._render();
  }
  init(): void {
    this.app.root.appendChild(this.pre);
  }
  destory(): void {
    this.app.root.removeChild(this.pre);
  }
  displayCode(code: string): void {
    this.code = code;
    this._render();
  }

  _render = () => {
    if (!this.selection) {
      this.pre.textContent = this.code;
      return;
    }

    const codeBeforeSelection = this.code.slice(0, this.selection.start);
    const codeSelected = this.code.slice(
      this.selection.start,
      this.selection.end,
    );
    const codeAfterSelection = this.code.slice(this.selection.end);

    const beforeEl = document.createElement("span");
    const selEl = document.createElement("mark");
    const afterEl = document.createElement("span");

    beforeEl.textContent = codeBeforeSelection;
    selEl.textContent = codeSelected;
    afterEl.textContent = codeAfterSelection;

    this.pre.replaceChildren(beforeEl, selEl, afterEl);
    selEl.scrollIntoView({
      block: "center",
    });
  };
}

export class GLApp {
  readonly mode: AppMode;
  readonly root: HTMLElement;

  readonly codeDisplay: CodeDisplay;

  renderer?: Renderer;

  audioAnalyser?: Analyser;

  animationFrame?: number;

  renderFunc: UserRenderFunction;

  renderFuncState: object;

  selectionBC: BroadcastChannel;

  midiMix: MIDIMix;

  constructor(mode: AppMode, root: HTMLElement, midiMix: MIDIMix) {
    this.mode = mode;
    this.root = root;

    this.midiMix = midiMix;

    if (this.mode !== "demo") {
      this.codeDisplay = new TextareaCodeDisplay(this);
    } else {
      this.codeDisplay = new DemoCodeDisplay(this);
    }

    this.selectionBC = new BroadcastChannel("jump2_textarea_selection");

    this.selectionBC.addEventListener("message", this.handleSelectionBroadcast);

    // set to a noop function
    this.renderFunc = () => {};

    // set state to an empty state
    this.renderFuncState = {};
  }

  updateCode = (code: string) => {
    this.compileCode(code);
    updateCodeInLocalStorage(code);
  };

  updateSelection = (selection: CursorSelection) => {
    this.selectionBC.postMessage(selection);
  };

  handleSelectionBroadcast = (e: MessageEvent) => {
    const { data } = e;
    if (!isCursorSelection(data)) {
      return;
    }
    this.codeDisplay.setSelection(data);
  };

  compileCode(code: string) {
    try {
      this.renderFunc = createFunction(code);
    } catch (e) {
      console.error(e);
      this.codeDisplay.setIsError(true);
      return;
    }

    this.codeDisplay.setIsError(false);
  }

  handleStorageEvent = (e: StorageEvent) => {
    if (e.storageArea !== window.localStorage) {
      return;
    }
    if (e.key !== codeStorageKey) {
      return;
    }
    const newCode = e.newValue;

    if (!newCode) {
      console.error("Where is the new code bruv?");
      return;
    }

    this.codeDisplay.displayCode(newCode);
    this.compileCode(newCode);
  };

  initializeLocalStoreageSync = () => {
    window.addEventListener("storage", this.handleStorageEvent);
  };

  setupRenderer(canvas: HTMLCanvasElement, gl: WebGL2RenderingContext) {
    this.renderer = new Renderer(canvas, gl, this.midiMix);
  }

  start() {
    const code = getStartingCode();
    this.codeDisplay.init();
    this.codeDisplay.displayCode(code);
    this.compileCode(code);

    this.initializeLocalStoreageSync();

    this.animationFrame = requestAnimationFrame(this._render);
  }

  stop() {
    if (this.animationFrame) {
      cancelAnimationFrame(this.animationFrame);
    }

    this.codeDisplay.destory();

    // clean up event listeners
    window.removeEventListener("storage", this.handleStorageEvent);
    this.selectionBC.removeEventListener(
      "message",
      this.handleSelectionBroadcast,
    );
    this.selectionBC.close();
  }

  setAudioAnalyser(audioAnalyser: Analyser) {
    this.audioAnalyser = audioAnalyser;
  }

  _render = (time: number) => {
    if (!this.renderer) {
      return;
    }

    this.renderer.update(time * 0.001);

    if (this.audioAnalyser) {
      this.renderer.updateFFTData(this.audioAnalyser.getFFTData());
      this.renderer.updateTempoData(this.audioAnalyser.updateBeat());
    }

    this.renderer.render((manager) => {
      this.renderFunc(this.renderFuncState, manager);
    });

    this.animationFrame = requestAnimationFrame(this._render);
  };
}
