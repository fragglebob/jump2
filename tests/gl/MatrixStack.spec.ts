import { m4 } from "twgl.js";
import { beforeEach, describe, expect, it } from "vitest";
import { MatrixStack } from "../../src/gl/MatrixStack";

describe("MatrixStack", () => {
	let stack: MatrixStack;

	beforeEach(() => {
		stack = new MatrixStack();
	});

	it("should start with a identity matrix", () => {
		expect(stack.current).toEqual(m4.identity());
	});

	it("push#should copy the matrix to the top of the stack when pushed", () => {
		const existing = stack.current;
		stack.push();
		expect(stack.current).not.toBe(existing);
		expect(stack.current).toEqual(existing);
	});

	it("pop#should return to the previous value after a pop", () => {
		const firstMatrix = stack.current;
		stack.push();
		const secondMatrix = stack.current;
		stack.pop();
		expect(stack.current).not.toBe(secondMatrix);
		expect(stack.current).toBe(firstMatrix);
	});

	it("pop#should allow popping with no depth, but it doesn't do anything", () => {
		const existing = stack.current;
		stack.pop();
		expect(stack.current).toBe(existing);
	});

	it("reset#should start again when calling reset", () => {
		const existing = stack.current;
		stack.reset();
		expect(stack.current).not.toBe(existing);
		expect(stack.current).toEqual(existing);
	});

	describe("transforms", () => {
		it("translate#should translate the current matrix", () => {
			const existing = stack.current;
			stack.translate([1, 2, 3]);
			expect(stack.current).toBe(existing);
			expect(stack.current).not.toEqual(m4.identity());
			expect(stack.current).toMatchSnapshot();
		});

		it("scale#should scale the current matrix", () => {
			const existing = stack.current;
			stack.scale([1, 2, 3]);
			expect(stack.current).toBe(existing);
			expect(stack.current).not.toEqual(m4.identity());
			expect(stack.current).toMatchSnapshot();
		});

		it("rotateX#should rotate the current matrix", () => {
			const existing = stack.current;
			stack.rotateX(123);
			expect(stack.current).toBe(existing);
			expect(stack.current).not.toEqual(m4.identity());
			expect(stack.current).toMatchSnapshot();
		});

		it("rotateY#should rotate the current matrix", () => {
			const existing = stack.current;
			stack.rotateY(123);
			expect(stack.current).toBe(existing);
			expect(stack.current).not.toEqual(m4.identity());
			expect(stack.current).toMatchSnapshot();
		});

		it("rotateZ#should rotate the current matrix", () => {
			const existing = stack.current;
			stack.rotateZ(123);
			expect(stack.current).toBe(existing);
			expect(stack.current).not.toEqual(m4.identity());
			expect(stack.current).toMatchSnapshot();
		});

		it("should undo transforms that happen in push", () => {
			stack.push();
			stack.translate([4.2, 1.23, 2.6]);
			stack.rotateZ(13.37);
			stack.scale([0.5, 2, 0.6656]);
			expect(stack.current).not.toEqual(m4.identity());
			stack.pop();
			expect(stack.current).toEqual(m4.identity());
		});
	});
});
