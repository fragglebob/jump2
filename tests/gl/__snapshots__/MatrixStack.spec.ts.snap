// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MatrixStack > transforms > rotateX#should rotate the current matrix 1`] = `
Float32Array [
  1,
  0,
  0,
  0,
  -0,
  -0.887968897819519,
  -0.4599034786224365,
  -0,
  0,
  0.4599034786224365,
  -0.887968897819519,
  0,
  0,
  0,
  0,
  1,
]
`;

exports[`MatrixStack > transforms > rotateY#should rotate the current matrix 1`] = `
Float32Array [
  -0.887968897819519,
  0,
  0.4599034786224365,
  0,
  0,
  1,
  0,
  0,
  -0.4599034786224365,
  -0,
  -0.887968897819519,
  -0,
  0,
  0,
  0,
  1,
]
`;

exports[`MatrixStack > transforms > rotateZ#should rotate the current matrix 1`] = `
Float32Array [
  -0.887968897819519,
  -0.4599034786224365,
  -0,
  -0,
  0.4599034786224365,
  -0.887968897819519,
  0,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  1,
]
`;

exports[`MatrixStack > transforms > scale#should scale the current matrix 1`] = `
Float32Array [
  1,
  0,
  0,
  0,
  0,
  2,
  0,
  0,
  0,
  0,
  3,
  0,
  0,
  0,
  0,
  1,
]
`;

exports[`MatrixStack > transforms > translate#should translate the current matrix 1`] = `
Float32Array [
  1,
  0,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  1,
  0,
  1,
  2,
  3,
  1,
]
`;
